### PR TITLE
[Bug] hs_persist_state drops indexed array elements (#3)

### LINKED ISSUE #3: [MAJOR] hs_persist_state drops indexed array elements
## Summary
Indexed arrays are persisted as scalars; only the first element survives, with no warning.

## Current behavior
- `local -a items=(one two)` persisted via `hs_persist_state items` emits a scalar assignment.
- The resulting state restores only `items[0]` and drops the rest.

## Expected behavior
- Either persist full indexed arrays, or explicitly error out when arrays are passed.

## Severity
Major (chosen as the easier of indexed/associative array support to implement first).

## Notes
- Associative arrays can be represented as two indexed arrays (keys/values), while indexed arrays can be represented as associative arrays as well.
- Variable-length structures are otherwise difficult to encode without string serialization.

### PR DESCRIPTION
**Pull request by lepeuvedic-natixar**
## Description

Implements full indexed-array (and associative-array) support in the HS2 state
format, fixing the silent data-loss reported in #3.

## Type of Change

- [x] Bug fix
- [x] New feature (array / nameref round-trip in HS2 format)
- [ ] Breaking change
- [x] Documentation update

### `feat(handle_state): implement HS2 state format with array and nameref support`
- Introduces the HS2 opaque state format (checksum-guarded, type-tagged records).
- `hs_persist_state` now emits full indexed-array and associative-array records
  instead of silently dropping all but the first element.
- Namerefs are persisted as `declare -n` records and round-trip correctly via the
  eval/stdout restore form.

### `refactor(handle_state): drop legacy code paths; allow namerefs in explicit restore`
- Removes all HS1 compatibility shims.
- Explicit restore form can now target a `declare -n` nameref declared in any
  ancestor scope.

### `fix(handle_state): all-or-nothing restore, trailing-var isolation, guard tests`
- `hs_read_persisted_state` explicit form: validates every guard condition
  (declared, not already set) for all requested names before any restoration
  occurs, making the operation atomic.
- `_hs_resolve_state_inputs`: fixes trailing-variable suffix isolation when
  known options (`-S`, `-q`) appear before the variable list; removes the
  now-unnecessary `$4` collision check.
- Docs: documents the all-or-nothing guarantee; corrects the
  `_hs_resolve_state_inputs` error description.

### Documentation and test commits
- `docs(handle_state)`: updated for HS2 format, `hs_persist_state` rename, and
  all new behavioral guarantees.
- `test(hs_persist_state)`: preliminary xfail tests, then full edge-case
  coverage — including pre-set/empty arrays and namerefs, all-or-nothing
  atomicity, ancestor-scope restore, and forwarded-arg handling.

## Testing

- [x] 82/82 bats tests pass (`bats test/test-hs_persist_state.bats`)
- [x] All edge cases covered: scalars, indexed arrays, associative arrays,
  namerefs, globals, all-or-nothing validation, corrupt state, forwarded args
- [x] Manual testing completed

## Related Issues

Closes #3
Related to #2 (function-name rejection — already closed)

## Checklist

- [x] Code follows project standards
- [x] Documentation updated (`docs/libraries/handle_state.rst`)
- [x] Tests pass (82/82)
- [x] Reviewed by team member

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: config/handle_state.sh [hunk 1]
```diff
@@ -47,145 +26,112 @@ readonly HS_ERR_MISSING_ARGUMENT=8
...
-        local status=$?
-        if [ $status -eq 124 ] || [ $status -eq 127 ] || [ $status -eq 137 ] || [ $status -eq 143 ]; then
-            echo "[ERROR] hs_persist_state_as_code: prior state is corrupted." >&2
-            return $((HS_ERR_CORRUPT_STATE))
-        elif [ $status -eq 1 ]; then
-            return $((HS_ERR_VAR_NAME_COLLISION))
-        elif [ $status -ne 0 ]; then
-            echo "[ERROR] hs_persist_state_as_code: internal error while checking for variable name collisions." >&2
-            return $((HS_ERR_CORRUPT_STATE))
+#   See hs_persist_state_as_code for the full error code list; additionally:
```
Comment (lepeuvedic) [inline config/handle_state.sh]: Obsolete reference to hs_persist_state_as_code, but the referenced material is gone, and must be restored.
Complete API function headers are required.
In addition, having a short usage example in the file header block is also necessary.

File: config/handle_state.sh [hunk 2]
```diff
@@ -47,145 +26,112 @@ readonly HS_ERR_MISSING_ARGUMENT=8
...
+        fi
+        __hsp_namerefs+=("$(_hs_strip_export "$__hsp_decl")")
+    done
+
+    # Build HS2 state: existing payload + non-nameref records + nameref records.
+    _hs_hs2_build "$__hsp_out_var" "$__hsp_existing_payload" \
+        "${__hsp_non_namerefs[@]}" "${__hsp_namerefs[@]}"
 }
 
 # --- hs_destroy_state ---------------------------------------------------------------
```
Comment (lepeuvedic) [inline config/handle_state.sh:165]: New hs_destroy_state code lacks the good inline comments the pre-HS2 code had.

File: docs/libraries/handle_state.rst [hunk 3]
```diff
@@ -281,15 +287,17 @@ Error Codes
...
 
 Known Limitations
 -----------------
 
-The tests currently demonstrate these limitations:
-
-- indexed arrays preserve only their first element
-- associative arrays are ignored
-- namerefs are restored as scalar values
+- Namerefs cannot be restored via the explicit restore form of
```
Comment (lepeuvedic) [inline docs/libraries/handle_state.rst]: If it is a stale comment, remove it. Otherwise if it describes reality, it's a bug that must be fixed.

**Review by lepeuvedic — CHANGES_REQUESTED** [review]
See comments attached to the files.

**Comment by lepeuvedic-natixar** [comment]
## Response to review comments

Three items addressed in commit `dc7ad04`:

**`config/handle_state.sh` line 44 — `hs_persist_state` header**
Replaced the stale `hs_persist_state_as_code` reference with the complete inline error-code list (eight entries), a full Options/Arguments section styled after `hs_read_persisted_state`, and two usage examples.

**`config/handle_state.sh` line 137 — `hs_destroy_state` body**
Added four phase comments inside the function body: parse + presence-map, validate, collect survivors, write-back.

**`docs/libraries/handle_state.rst` line 295 — stale nameref limitation**
The limitation was accurate for HS1 but became stale after commit `c2d753e` ("allow namerefs in explicit restore"). Removed from both the Explicit restore Behavior bullet list and the Known Limitations section. The existing tests (`hs_read_persisted_state explicit form restores nameref and target in ancestor scope`) confirm the explicit form works for namerefs.

All 82 tests still pass.

**Review by lepeuvedic — APPROVED** [review]
**Approved**
